File: sdks/markets/rust/src/ix/funding.rs
//! Funding-phase instruction builders (Ix 2 `CreateMarket`, Ix 3 `Contribute`,
//! Ix 4 `Cancel`, Ix 5 `Refund`).

use crate::*;
use solana_sdk::instruction::{AccountMeta, Instruction};
use solana_sdk::pubkey::Pubkey;

/// `CreateMarket` (Ix 2) — create the `outcome_index` binary sub-market for
/// `oracle`, its KASS escrow, and the creator's `Contribution`, transferring
/// `seed_amount` KASS in.
/// Payload = `seed_amount` (u64 LE) ++ `outcome_index` (u8). Accounts:
/// `[0] config(ro) [1] oracle(ro) [2] market(pda,w) [3] escrow(pda,w)
///  [4] kass_mint(ro) [5] creator(signer,w) [6] creator_kass_ata(w)
///  [7] contribution(pda,w) [8] token program [9] system program`.
#[allow(clippy::too_many_arguments)]
pub fn create_market(
    creator: &Pubkey,
    oracle: &Pubkey,
    kass_mint: &Pubkey,
    creator_kass_ata: &Pubkey,
    seed_amount: u64,
    outcome_index: u8,
) -> Instruction {
    let (config, _) = crate::pda::config();
    let (market, _) = crate::pda::market(oracle, outcome_index);
    let (escrow, _) = crate::pda::escrow(&market);
    let (contribution, _) = crate::pda::contribution(&market, creator);
    let mut data = vec![IX_CREATE_MARKET];
    data.extend_from_slice(&seed_amount.to_le_bytes());
    data.push(outcome_index);
    Instruction {
        program_id: PROGRAM_ID,
        accounts: vec![
            AccountMeta::new_readonly(config, false),
            AccountMeta::new_readonly(*oracle, false),
            AccountMeta::new(market, false),
            AccountMeta::new(escrow, false),
            AccountMeta::new_readonly(*kass_mint, false),
            AccountMeta::new(*creator, true),
            AccountMeta::new(*creator_kass_ata, false),
            AccountMeta::new(contribution, false),
            AccountMeta::new_readonly(spl_token::id(), false),
            AccountMeta::new_readonly(solana_sdk::system_program::id(), false),
        ],
        data,
    }
}

/// `Contribute` (Ix 3) — add `amount` KASS to a `Funding` market's escrow and
/// create-or-increment the contributor's `Contribution`.
/// Payload = `amount` (u64 LE). Accounts:
/// `[0] market(w) [1] escrow(w) [2] contributor(signer,w) [3] contributor_kass_ata(w)
///  [4] contribution(pda,w) [5] token program [6] system program`.
///
/// The processor reads only the first six accounts (its slice pattern tolerates
/// trailing accounts). The system program is appended so it is loaded into the
/// transaction: the first-ever contribution from a given contributor creates
/// their `Contribution` PDA via a CPI to the system program's `CreateAccount`,
/// which requires that program to be present in the tx account set (same as
/// `create_market`, which passes it explicitly).
pub fn contribute(
    contributor: &Pubkey,
    market: &Pubkey,
    escrow: &Pubkey,
    contributor_ata: &Pubkey,
    amount: u64,
) -> Instruction {
    let (contribution, _) = crate::pda::contribution(market, contributor);
    let mut data = vec![IX_CONTRIBUTE];
    data.extend_from_slice(&amount.to_le_bytes());
    Instruction {
        program_id: PROGRAM_ID,
        accounts: vec![
            AccountMeta::new(*market, false),
            AccountMeta::new(*escrow, false),
            AccountMeta::new(*contributor, true),
            AccountMeta::new(*contributor_ata, false),
            AccountMeta::new(contribution, false),
            AccountMeta::new_readonly(spl_token::id(), false),
            AccountMeta::new_readonly(solana_sdk::system_program::id(), false),
        ],
        data,
    }
}

/// `Cancel` (Ix 4) — mark an under-funded `Funding` market `Cancelled` once its
/// underlying Kassandra oracle is terminal. Permissionless (no required signer
/// beyond the tx fee payer). Payload = empty. Accounts:
/// `[0] market(w) [1] oracle(ro)`.
pub fn cancel(market: &Pubkey, oracle: &Pubkey) -> Instruction {
    Instruction {
        program_id: PROGRAM_ID,
        accounts: vec![
            AccountMeta::new(*market, false),
            AccountMeta::new_readonly(*oracle, false),
        ],
        data: vec![IX_CANCEL],
    }
}

/// `Refund` (Ix 5) — permissionless per-contributor refund from a `Cancelled`
/// market. Program-signed transfer of the recorded stake out of escrow back to
/// the contributor's KASS ata, then the `Contribution` is CLOSED with its rent
/// returned to `contributor`. Payload = empty. Accounts:
/// `[0] market(w) [1] escrow(w) [2] contribution(w) [3] contributor_kass_ata(w)
///  [4] contributor(w) [5] token program`.
///
/// `market` is writable (its `open_contributions` counter is decremented) and
/// `contributor` (== `contribution.contributor`) receives the closed Contribution's
/// rent.
pub fn refund(
    market: &Pubkey,
    escrow: &Pubkey,
    contribution: &Pubkey,
    contributor_ata: &Pubkey,
    contributor: &Pubkey,
) -> Instruction {
    Instruction {
        program_id: PROGRAM_ID,
        accounts: vec![
            AccountMeta::new(*market, false),
            AccountMeta::new(*escrow, false),
            AccountMeta::new(*contribution, false),
            AccountMeta::new(*contributor_ata, false),
            AccountMeta::new(*contributor, false),
            AccountMeta::new_readonly(spl_token::id(), false),
        ],
        data: vec![IX_REFUND],
    }
}

File: sdks/markets/ts/src/instructions/market/funding.ts
/**
 * Funding-phase instruction builders (Ix 2–5): create / contribute / cancel / refund.
 *
 * See `../market/index.js` for the module overview. Account orders + payload
 * layouts are mirrored VERBATIM from the verified Rust builders in
 * `sdks/oracles/rust/src/ix.rs` (a mismatch is a silent runtime failure).
 */
import { Address, TransactionInstruction } from "@solana/web3.js";

import { Ix, MARKET_PROGRAM_ID, SYSTEM_PROGRAM_ID, TOKEN_PROGRAM_ID } from "../../constants.js";
import * as pda from "../../pda.js";
import type { AddressInput } from "../../pda.js";
import { addr, ro, u64LE, u8, w, withDisc } from "../payload.js";

// ---------------------------------------------------------------------------
// CreateMarket (Ix 2) — create the `outcome_index` binary sub-market for `oracle`,
// its KASS escrow, and the creator's Contribution, transferring `seed_amount`
// KASS in. Binary markets pass `outcomeIndex = 0`; a categorical oracle has one
// sub-market per outcome (the market PDA is keyed by `(oracle, outcomeIndex)`).
// Payload = seed_amount(u64 LE) ++ outcome_index(u8).
// Accounts: 0 config(w) 1 oracle(ro) 2 market(w,PDA) 3 escrow(w,PDA)
//           4 kass_mint(ro) 5 creator(signer,w) 6 creator_kass_ata(w)
//           7 contribution(w,PDA) 8 token program(ro) 9 system program(ro).
// `config` is WRITABLE: create_market bumps its market-creation-activity EMA
// (see the program's `liquidity_floor` module).
// ---------------------------------------------------------------------------
export interface CreateMarketArgs {
  /** Creator (signer): pays rent + seeds the first contribution. */
  creator: AddressInput;
  /** The Kassandra oracle the market resolves against (seeds the market PDA). */
  oracle: AddressInput;
  /** Canonical KASS mint (== `config.kass_mint`). */
  kassMint: AddressInput;
  /** Creator's KASS token account the seed amount transfers from. */
  creatorKassAta: AddressInput;
  /** KASS seeded into escrow as the creator's contribution. */
  seedAmount: bigint | number;
  /**
   * The oracle outcome this sub-market binds to (`0 <= outcomeIndex <
   * oracle.options_count`); YES = the oracle resolves to this index. Binary
   * markets pass `0`. Keys the market/escrow/contribution PDAs.
   */
  outcomeIndex: number;
  programId?: Address;
}

export async function createMarket(args: CreateMarketArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? MARKET_PROGRAM_ID;
  const config = await pda.config(programId);
  const market = await pda.market(args.oracle, args.outcomeIndex, programId);
  const escrow = await pda.escrow(market.address, programId);
  const contribution = await pda.contribution(market.address, args.creator, programId);
  return new TransactionInstruction({
    programId,
    keys: [
      w(config.address),
      ro(addr(args.oracle)),
      w(market.address),
      w(escrow.address),
      ro(addr(args.kassMint)),
      w(addr(args.creator), true),
      w(addr(args.creatorKassAta)),
      w(contribution.address),
      ro(TOKEN_PROGRAM_ID),
      ro(SYSTEM_PROGRAM_ID),
    ],
    data: withDisc(Ix.CreateMarket, u64LE(args.seedAmount), u8(args.outcomeIndex)),
  });
}

// ---------------------------------------------------------------------------
// Contribute (Ix 3) — add `amount` KASS to a Funding market's escrow and
// create-or-increment the contributor's Contribution.
// Payload = amount(u64 LE).
// Accounts: 0 market(w) 1 escrow(w,PDA) 2 contributor(signer,w) 3 contributor_kass_ata(w)
//           4 contribution(w,PDA) 5 token program(ro) 6 system program(ro).
// ---------------------------------------------------------------------------
export interface ContributeArgs {
  /** Contributor (signer): funds the stake + rent for a first-time Contribution. */
  contributor: AddressInput;
  /** The market being contributed to. */
  market: AddressInput;
  /** Contributor's KASS token account the stake transfers from. */
  contributorKassAta: AddressInput;
  /** KASS to stake (raw base units). */
  amount: bigint | number;
  programId?: Address;
}

export async function contribute(args: ContributeArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? MARKET_PROGRAM_ID;
  const escrow = await pda.escrow(args.market, programId);
  const contribution = await pda.contribution(args.market, args.contributor, programId);
  return new TransactionInstruction({
    programId,
    keys: [
      w(addr(args.market)),
      w(escrow.address),
      w(addr(args.contributor), true),
      w(addr(args.contributorKassAta)),
      w(contribution.address),
      ro(TOKEN_PROGRAM_ID),
      ro(SYSTEM_PROGRAM_ID),
    ],
    data: withDisc(Ix.Contribute, u64LE(args.amount)),
  });
}

// ---------------------------------------------------------------------------
// Cancel (Ix 4) — mark an under-funded Funding market Cancelled once its oracle
// is terminal. Permissionless. Payload = empty.
// Accounts: 0 market(w) 1 oracle(ro).
// ---------------------------------------------------------------------------
export interface CancelArgs {
  /** The market to cancel. */
  market: AddressInput;
  /** The market's Kassandra oracle (must be terminal). */
  oracle: AddressInput;
  programId?: Address;
}

export async function cancel(args: CancelArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? MARKET_PROGRAM_ID;
  return new TransactionInstruction({
    programId,
    keys: [w(addr(args.market)), ro(addr(args.oracle))],
    data: withDisc(Ix.Cancel),
  });
}

// ---------------------------------------------------------------------------
// Refund (Ix 5) — permissionless per-contributor refund from a Cancelled market.
// The Contribution is CLOSED (rent → contributor). Payload = empty.
// Accounts: 0 market(w) 1 escrow(w,PDA) 2 contribution(w,PDA) 3 contributor_kass_ata(w)
//           4 contributor(w) 5 token program(ro).
// `market` is writable (its open_contributions counter is decremented) and
// `contributor` (== contribution.contributor) receives the closed Contribution's rent.
// ---------------------------------------------------------------------------
export interface RefundArgs {
  /** The Cancelled market (writable — its open_contributions counter decrements). */
  market: AddressInput;
  /**
   * The contributor being refunded (seeds the Contribution PDA AND is the rent
   * recipient of the closed Contribution — must equal `contribution.contributor`).
   */
  contributor: AddressInput;
  /** Contributor's KASS token account the stake refunds to. */
  contributorKassAta: AddressInput;
  programId?: Address;
}

export async function refund(args: RefundArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? MARKET_PROGRAM_ID;
  const escrow = await pda.escrow(args.market, programId);
  const contribution = await pda.contribution(args.market, args.contributor, programId);
  return new TransactionInstruction({
    programId,
    keys: [
      w(addr(args.market)),
      w(escrow.address),
      w(contribution.address),
      w(addr(args.contributorKassAta)),
      w(addr(args.contributor)),
      ro(TOKEN_PROGRAM_ID),
    ],
    data: withDisc(Ix.Refund),
  });
}
